### PR TITLE
Create the docs for the MPU API

### PR DESCRIPTION
**Pull request by c1728p9**
Add the HAL spec porting guide for the MPU API.

### CODE REVIEW

File: docs/porting/target/mpu.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+To enable MPU support in Mbed OS, add the `MPU` label in the `device_has` option of the target's section in the `targets.json` file.
+
+Targets with a standard ARMv7-M or ARMv8-M MPU, indicated by `__MPU_PRESENT` being defined to 1 in the target's CMSIS header, only need the `MPU` label in the `device_has` for MPU support. This pulls in a common Mbed OS MPU driver, so in this case a target specific driver is not needed.
+
+Targets with a standard ARMv7-M or ARMv8-M MPU needing to override the common Mbed OS MPU driver can do so by defining `MBED_MPU_CUSTOM`. This removes the common Mbed OS MPU driver from builds so a target specific one can be used instead.
+
+Targets with a standard ARMv7-M or ARMv8-M MPU with the common Mbed OS MPU driver can customize where flash ends and ram begins by defining `MBED_MPU_ROM_END`. By default this is set to `0x0FFFFFFF` for  ARMv7-M and `0x1FFFFFFF` for ARMv8-M. The address space up to and including this is treated as rom which the MPU protects from write accesses. The address space after this is treated as ram which the MPU protects from code execution.
```
Comment (c1728p9) [inline docs/porting/target/mpu.md]: This paragraph needs to be removed since a separate mechanism will be used for this

**Comment by AnotherButler** [comment]
ping @donatieng @scartmell-arm

File: docs/api/api.md [hunk 2]
```diff
@@ -32,6 +32,11 @@
 <td><a href="rtc.html">RTC</a></td>
 <td><a href="assert.html">Assert</a></td>
 <td><a href="platformmutex.html">PlatformMutex</a></td>
+<td><a href="mpu-management.html">MPU management</a></td>
+</tr>
+<tr>
+<td><a href="scopedramexecutionlock.html">ScopedRamExecutionLock</a></td>
+<td><a href="scopedromwritelock.html">ScopedRomWriteLock</a></td>
```
Comment (AnotherButler) [inline docs/api/api.md]: The tables go in order from top to bottom, not left to right. If you don't want to fix this, I can do so after the PR goes in.

**Comment by AnotherButler** [comment]
This is waiting on engineering review and the code to go on master: https://github.com/ARMmbed/mbed-os/pull/8335/ (also referenced above)

File: docs/api/platform/MpuManagement.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+- Prevents execution from RAM.
+- Prevents writing to ROM.
+
+Mbed OS handles MPU management automatically in the following situations:
+
+- Memory protection is enabled as part of the boot sequence.
+- Memory protection is disabled when starting a new application.
+- Memory protection is disabled while flash programming.
+
+<span class="notes">**Note:** Memory protection should be transparent to most applications and libraries because Mbed OS handles it automatically for operations that need to disable MPU protections, such as flash programming.</span>
```
Comment (donatieng) [inline docs/api/platform/MpuManagement.md]: Maybe add "this is an advanced feature, most users should not have to worry about this - only use it if you know what you're doing" type of warning?

File: docs/api/platform/MpuManagement.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+#### RAM execute lock
+
+After-boot execution from RAM is not allowed. Libraries requiring the ability to execute from RAM can enable this by acquiring the RAM execution lock. The RAM execution lock has a count associated with it, and you can lock t multiple times. Execution from RAM is disabled only when all components have unlocked it.
```
Comment (donatieng) [inline docs/api/platform/MpuManagement.md]: After-boot --> After boot?

Comment (donatieng) [reply docs/api/platform/MpuManagement.md]: you can lock t --> it

File: docs/api/platform/MpuManagement.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+### Example
+
+```C
+#include "mbed.h"
+
+int main()
+{
+    // Permanently enable execute from RAM for this application
+    mbed_mpu_manager_lock_mem_xn();
```
Comment (donatieng) [inline docs/api/platform/MpuManagement.md]: Can we push the C++ RAII APIs instead as user-facing APIs?

File: docs/porting/target/mpu.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+### Dependency
+
+Hardware MPU capabilities.
+
+### Implementing the MPU API
+
+You can find the API and specification for the MPU API in the following header file:
+
+[![View code](https://www.mbed.com/embed/?type=library)](http://os-doc-builder.test.mbed.com/docs/development/mbed-os-api-doxy/group__hal__mpu.html)
```
[the remark this answers is not in the document]
Comment (c1728p9) [reply docs/porting/target/mpu.md]: Nope, it hasn't yet.

File: docs/api/platform/MpuManagement.md (hunk 5 above)
Comment (c1728p9) [reply docs/api/platform/MpuManagement.md]: done

File: docs/api/platform/MpuManagement.md (hunk 4 above)
Comment (c1728p9) [reply docs/api/platform/MpuManagement.md]: done

File: docs/api/platform/MpuManagement.md (hunk 3 above)
Comment (c1728p9) [reply docs/api/platform/MpuManagement.md]: @AnotherButler had concerning about the phrasing here. I'll let her make this change.

File: docs/api/api.md (hunk 2 above)
Comment (c1728p9) [reply docs/api/api.md]: done

**Comment by AnotherButler** [comment]
@donatieng I believe we've addressed all of your comments. Please rereview.

**Comment by AnotherButler** [comment]
@donatieng Could you please rereview?

**Pull request merged**